File: cses/01_introductory/1092_two-sets/solution.py
import sys


def main():
    data = sys.stdin.buffer.read().split()
    n = int(data[0])

    # Tổng 1..n = n*(n+1)/2 chỉ chẵn khi n % 4 == 0 hoặc n % 4 == 3; ngoài ra không chia được.
    if n % 4 != 0 and n % 4 != 3:
        sys.stdout.write("NO\n")
        return

    target = n * (n + 1) // 4  # tổng mà mỗi tập cần đạt được

    set1 = []
    set2 = []
    remaining = target
    # Tham lam: duyệt số lớn nhất về nhỏ nhất, lấy i vào tập 1 khi còn "vừa" (i <= remaining),
    # phần còn lại dồn sang tập 2 nên tổng tập 2 cũng bằng target.
    for i in range(n, 0, -1):
        if i <= remaining:
            set1.append(i)
            remaining -= i
        else:
            set2.append(i)

    # Gom kết quả rồi xuất một lần để tối ưu I/O.
    out = []
    out.append("YES")
    out.append(str(len(set1)))
    out.append(" ".join(map(str, set1)))
    out.append(str(len(set2)))
    out.append(" ".join(map(str, set2)))
    sys.stdout.write("\n".join(out) + "\n")


if __name__ == "__main__":
    main()

File: cses/01_introductory/1092_two-sets/solution.cpp
#include <iostream>
#include <vector>

using namespace std;

// In một tập: dòng đầu là số phần tử, dòng sau là các phần tử cách nhau bởi dấu cách.
void print_set(const vector<int>& values) {
    cout << values.size() << '\n';
    for (size_t i = 0; i < values.size(); ++i) {
        if (i > 0) {
            cout << ' ';
        }
        cout << values[i];
    }
    cout << '\n';
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    cin >> n;

    // Tổng 1..n = n*(n+1)/2 chỉ chẵn khi n % 4 == 0 hoặc n % 4 == 3; ngoài ra không chia được.
    if (n % 4 != 0 && n % 4 != 3) {
        cout << "NO\n";
        return 0;
    }

    // Mỗi tập cần đạt tổng target = S/2 = n*(n+1)/4.
    const long long target = 1LL * n * (n + 1) / 4;
    long long remaining = target;
    vector<int> first_set;
    vector<int> second_set;

    // Tham lam: duyệt số lớn nhất về nhỏ nhất, lấy i vào tập 1 khi còn "vừa" (i <= remaining),
    // phần còn lại dồn sang tập 2 nên tổng tập 2 cũng bằng target.
    for (int value = n; value >= 1; --value) {
        if (value <= remaining) {
            first_set.push_back(value);
            remaining -= value;
        } else {
            second_set.push_back(value);
        }
    }

    cout << "YES\n";
    print_set(first_set);
    print_set(second_set);
    return 0;
}
